x-pack/platform/plugins/shared/inference/server/chat_complete/anonymization/anonymize_messages.ts

```diff
@@ -0,0 +1,234 @@
+import objectHash from 'object-hash';
+import { ElasticsearchClient } from '@kbn/core/server';
+import type { Logger } from '@kbn/logging';
+import pLimit from 'p-limit';
+import { getAnonymizableMessageParts } from './get_anonymizable_message_parts';
+
+const DEFAULT_MAX_CONCURRENT_REQUESTS = 5;
+const DEFAULT_BATCH_SIZE = 1_000;
+const DEFAULT_CHUNK_SIZE = 1_000;
+
+export interface InferenceChunk {
+  chunkText: string;
+  charStartOffset: number;
+}
+
+export function chunkText(text: string, maxChars = DEFAULT_CHUNK_SIZE): InferenceChunk[] {
+  const chunks: InferenceChunk[] = [];
+  for (let i = 0; i < text.length; i += maxChars) {
+    chunks.push({
+      chunkText: text.slice(i, i + maxChars),
+      charStartOffset: i,
+    });
+  }
+  return chunks;
+}
+
+function getEntityMask(entity: { class_name: string; value: string }) {
+  const hash = objectHash({
+    value: entity.value,
+    class_name: entity.class_name,
+  });
+  return `${entity.class_name}_${hash}`;
+}
+
+export async function anonymize({
+  input,
+  anonymizationRules,
+  esClient,
+}: {
+  input: string;
+  anonymizationRules: AnonymizationRule[];
+  esClient: ElasticsearchClient;
+}): Promise<{ output: string; anonymizations: Anonymization[] }> {
+  const [regexRules, nerRules] = partition(
+    anonymizationRules,
+    (rule): rule is RegexAnonymizationRule => rule.type === 'RegExp'
+  );
+
+  const entities: Array<{ rule: RegexAnonymizationRule; entity: AnonymizationEntity }> = [];
+
+  let output = input;
+
+  regexRules.forEach((rule) => {
+    const regex = new RegExp(rule.pattern);
+
+    let match: RegExpMatchArray | null = null;
+    while ((match = regex.exec(output))) {
+      const value = match[0];
+
+      const mask = getEntityMask({
+        value,
+        class_name: rule.entityClass,
+      });
+
+      output = output.replace(match[0], mask);
+
+      entities.push({
+        entity: {
+          value,
+          class_name: rule.entityClass,
+          mask,
+        },
+        rule,
+      });
+    }
+  });
+
+  let index = 0;
+
+  const anonymizations: Anonymization[] = [];
+  entities.forEach(({ entity, rule }) => {
+    const start = output.indexOf(entity.mask, index);
+    index = start + entity.mask.length;
+    anonymizations.push({
+      entity,
+      rule: {
+        type: rule.type,
+      },
+    });
+  });
+
+  if (!nerRules.length) {
+    return {
+      output,
+      anonymizations,
+    };
+  }
+
+  // Maximum number of concurrent requests
+  const limiter = pLimit(DEFAULT_MAX_CONCURRENT_REQUESTS);
+
+  for (const nerRule of nerRules) {
+    let offset = 0;
+    // Chunk the text due to token limit of model
+    const chunks = chunkText(output);
+    // Limit chunks into batches per request
+    const batches = chunk(chunks, DEFAULT_BATCH_SIZE);
+
+    for (let batchIndex = 0; batchIndex < batches.length; batchIndex++) {
+      const batchChunks = batches[batchIndex];
+
+      const results = await limiter(async () =>
```

Comment:
limiter here doesn't actually make sense. it will only run one request this way anyway. I think it should be converted into a Promise.all()?